### ISSUE TITLE
add support for generator files in EDMLHE format

### ISSUE THREAD

**Issue by belforte** [dmwm/CRABServer#4822]
pasting here from HN:
On 06/05/2015 10:52 PM, Dinko Ferencek wrote:

Dear CRAB3 developers,

I have a CRAB3 feature request which hopefully should not be too 
difficult to implement. Basically, in addition to having the possibility 
to read LHE files using the LHESource when producing private MC, it 
would be very useful, and more efficient, to have a possibility to read 
LHE files stored in the EDM format (essentially, use as input LHE files 
that have been pre-processed using the LHESource, which is a very fast 
and can be easily done locally). In this way CRAB jobs would access ROOT 
files, which are compressed by default and should require less 
bandwidth, and skipping events when using multiple input files should be 
a lot more efficient. From what I've seen, LHESource seems to cycle 
through all previous events until it finds the one from which to start 
processing. In addition, until 7_1_16_patch2, LHESource did not properly 
support multiple input files.

Would this type of "JobType.generator = 'edmlhe'" feature be possible?

Many thanks,
Dinko

-------------------------------------------------------------
Visit this CMS message (to reply or unsubscribe) at: 
https://hypernews.cern.ch/HyperNews/CMS/get/computing-tools/667.html

**Comment by matz-e** [comment]
I am fairly sure that in this case, the MC generation workflow in CRAB can just be replaced by a regular analysis-type job.

In general, I think the labels of MC production and analysis are a little mis-leading, as the more refer to the input source of the data.  I.e., analysis workflows run on data in DBS / EDM format, whereas MC production has to produce data first.  Since the data in question is in EDM format, I'd try an analysis workflow.

**Comment by matz-e** [comment]
Come to think of it, for private input data, we are currently lacking some data discovery to use lumi-based splitting, which would make sense here (since the EDM files with LHE content are probably quite large/contain many events.)

**Comment by belforte** [comment]
I was also thinking that this may be an "analysis" workflow. But indeed 
the input may not be in DBS, besides not having ready metadata. Also I 
do not know if a special input is needed from CMSSW or normal EDM events 
can be used. On the other hand we support "run on a bunch of files not 
in DBS" IIUC, so the lack of DBS info may be OK.
Sure, if the issue is resolved with documentation, rather than code, all 
the better !

**Comment by ferencek** [comment]
Let me just add that I tried using
```python
config.JobType.pluginName = 'PrivateMC'
#config.JobType.generator = 'lhe'
```
hoping this was default to the standard PoolSource but my jobs were then failing with
```
== CMSSW:  ----- Begin Fatal Exception 04-Jun-2015 20:26:37 CEST-----------------------
== CMSSW:  An exception of category 'Configuration' occurred while
== CMSSW:     [0] Constructing the EventProcessor
== CMSSW:     [1] Validating configuration of input source of type PoolSource
== CMSSW:  Exception Message:
== CMSSW:  Illegal parameter found in configuration.  The parameter is named:
== CMSSW:   'numberEventsInLuminosityBlock'
== CMSSW:  You could be trying to use a parameter name that is not
== CMSSW:  allowed for this plugin or it could be misspelled.
== CMSSW:  ----- End Fatal Exception -------------------------------------------------
```
So it seems that CRAB in the 'PrivateMC' mode tries to pass to the PoolSource an illegal parameter since it is probably expecting an EmptySource.

**Comment by ferencek** [comment]
Concerning an analysis-type job, I agree, this is all that is principle needed. However, how is one supposed to tell CRAB not to look for data in DBS but just to use files included in the PoolSource of the CMSSW cfg file?

**Comment by ferencek** [comment]
OK, so I tried something like
```python
...
config.JobType.pluginName = 'Analysis'
#config.JobType.generator = 'lhe'
config.JobType.psetName = 'step0_Stop2ToStop1H_TuneCUETP8M1_13TeV-madgraph-pythia8_GEN.py'

config.Data.primaryDataset = 'Stop2ToStop1H_TuneCUETP8M1_13TeV-madgraph-pythia8'
config.Data.splitting = 'EventBased'
...
```
but then I get
```
Will use CRAB configuration file crab3_Stop2ToStop1H_TuneCUETP8M1_13TeV-madgraph-pythia8_GEN_step0.py
Invalid CRAB configuration: Analysis job type requires an input dataset or a set of user input files to run on.
Suggestion: To specify an input dataset use the parameter Data.inputDataset. To specify a set of user input files use the parameter Data.userInputFiles.
The documentation about the CRAB configuration file can be found in https://twiki.cern.ch/twiki/bin/view/CMSPublic/CRAB3ConfigurationFile.
```
so this does not yet seem to work.

**Comment by PerilousApricot** [comment]
Did you try ```Data.inputDataset```?

**Comment by PerilousApricot** [comment]
bah, ```Data.userInputFiles``` is what I meant...

**Comment by ferencek** [comment]
Yes, I just did. However, reading about `userInputFiles` at https://twiki.cern.ch/twiki/bin/view/CMSPublic/CRAB3ConfigurationFile, I see that the only allowed splitting mode is `FileBased`, which is not optimal (LHE files in EDM format can easily contain tens of thousands of events which would not be possible to run through the detector simulation in one job in a reasonable amount of time).

**Comment by PerilousApricot** [comment]
Right, but as an interim measure, you can split your LHE files up while the
necessary changes are meant. I think what you're looking for would need a
change on the server side, so it'd be a while before it could be
implemented (CRAB has to wait on the HTTP group for changes relating to
input parameters since the frontend is maintained by them)

**Comment by ferencek** [comment]
In principle yes, but having thousands of input EDM files is not a good solution either. I would say it is better to wait and just use the LHESource in the meantime ;)

I apologise on my ignorance, but why would this feature require any new input parameters. Wouldn't it be just a matter of how the existing ones are interpreted. In fact, the only change needed seem to be to allow event- and/or lumi-based based job splitting.

**Comment by belforte** [comment]
How can CRAB split by anything  but file, if all that it is given is a list of file names ?

That's why we usually start from datasets described in DBS where needed metadata like #events and/or #lumis in a file are kept.

You are basically making the case for providing splitting information to CRAB via some configuration information, which will always be very ad-hoc. Many things could be done, but why not turn the input list into a DBS dataset instead ? Seems the most general and least development effort solution to me. Maybe you can run a CRAB jobs with the current LHESource, produce the EDM files and publish them ? As you said this will be very fast and a few jobs will do.

**Comment by matz-e** [comment]
IMO, that involves another step in the workflow relying on the grid (and thus taking a random amount of time.)  I don't think it should be too hard to run `edmFileUtil` over the user-provided input files and send the data-discovery along with the job submission?

**Comment by belforte** [comment]
why not run emdFileUtil, prepare proper metadata, and store them in DBS instead ?
Publishing a dataset made of a handful of files is fast and simple.
Input files should be in /store/.... but I do not think this should be really
limiting. Or does it?

If nothing else, going via DBS specifies a metadata format that exiting CRAB code will understand.
Otherwise we (or the user) commits to always reproduce DBS output format, or (even worse)
internal CRAB objects representation, which surely we want to be free of changing in the long run.
The object formats inside CRAB may stay stable for years, but internal data structures should better 
stay private.

**Comment by belforte** [comment]
we discussed this at some length in today's CRAB3 development meeting.
No vote was taken, but a majority favored to keep client thin and configuration options limited. If nothing else to limit the possibility for malformed inputs resulting in subtle errors which causes expert to loose time on what often are simple mistakes by unexperienced users.

Current plan is thus "make it easy for users to put their LHE files in DBS, in particular in EDM format".

Details should be in more ad-hoc issues. But as AFAIU the work will go along a two-pronged approach:
1. finalise the famous "script" (see User Data Review) to allow user to insert a bunch of files in DBS. Stefano already has a script which can publish LHE files,taking #events/file from outside.
2. preferred still is that user can run LHE to EDM conversion in CRAB and publish the output. LHEsource has the problem that it can only deal with a list of 3 files, but:
   a) Matthias (?) fixed this in 7_5, could ask for a backport to 7_4
   b) for LHE to EDM only, cmsRun can be fast enough that split-by-file can do

**Comment by ericvaandering** [comment]
Let's also not forget the provenance tracking that having things published in DBS gives. The further back and forward we can extend that and our other tools, the better. (provided things are ~just as easy for the user.)

**Comment by ferencek** [comment]
Having the possibility to just publish a bunch of LHE files in the EDM format in DBS would be great. I agree that one can in principle run the LHE->EDM step via CRAB but this makes the entire step unnecessarily dependent on the grid.

The main reason I was asking for the option of using LHE files in the EDM format for a private MC production was because the LHESource seemed broken with more than 1 input file. Luckily, this got fixed in the last few days (well, the fix in https://github.com/cms-sw/cmssw/pull/9448 was apparently for another issue with more than 3 input files but from my experience things were broken even for more than 1 file). So in some sense, with this fix included, the case for using EDM files as input in a private MC production is a bit less critical. Nevertheless, I still think it is quite wasteful to access uncompressed LHE files via xrootd and it would be much better and more efficient to access files in the EDM format. What seems to be particularly wasteful is when the last job needs to open all LHE files one by one and cycle through all the events just to process a handful of events at the end of the last input file. With EDM files this event cycling would be avoided and there would a lot less network traffic generated.

**Comment by belforte** [comment]
I understand that  grid jobs take more time, and expose to possible delays, yet most of the time they work well and fast (in particular if a short maxWallTime is requested).
But surely I can be suspected of a biased opinion here.

In the end, about the "don't force me to go grid for a few, short jobs"  I'd rather look into giving users a way to run CRAB jobs interactively/locally (which may be good for other purposes as well and we are likely going to do anyhow e.g. for CAF), but keep the concept that the best place for user private MC production is files in /store/user/... and datasets in DBS. I.e. use the machine you run crab on as a WN.

I have the prejudice that the more we can user general/generic approach, rather then code around the specific of a particular workflow, the more useful our tools become.

**Comment by ferencek** [comment]
@belforte, being able to run CRAB jobs locally for the LHE->EDM step would be perfect. I also fully agree that it's best not to deviate from keeping files in /store/user/... and datasets in DBS. Thanks.

**Comment by belforte** [comment]
of course... do not hold your breath !

**Comment by mmascher** [comment]
Use case would be addressed by https://github.com/dmwm/CRABServer/issues/5140
